File: python/510.py
"""Project Euler Problem 510: Tangent Circles.

Find the sum of all r_A + r_B + r_C for all r_A ≤ r_B ≤ N such that two
circles with radii r_A and r_B are tangent to each other and a line L,
and a circle with radii r_C is internally tangent to all three.
"""

from __future__ import annotations

from math import isqrt
from typing import Dict, List, Set


def sieve_primes(limit: int) -> List[int]:
    """Sieve of Eratosthenes."""
    if limit < 2:
        return []
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False
    for i in range(2, isqrt(limit) + 1):
        if is_prime[i]:
            for j in range(i * i, limit + 1, i):
                is_prime[j] = False
    return [i for i in range(limit + 1) if is_prime[i]]


def prime_factorization(n: int, primes: List[int]) -> Dict[int, int]:
    """Prime factorization."""
    factors: Dict[int, int] = {}
    for p in primes:
        if p * p > n:
            break
        while n % p == 0:
            factors[p] = factors.get(p, 0) + 1
            n //= p
    if n > 1:
        factors[n] = factors.get(n, 0) + 1
    return factors


def all_divisors(n: int, prime_factors: Set[int]) -> List[int]:
    """Get all divisors of n using prime factors."""
    divisors = [1]
    for p in prime_factors:
        new_divisors = []
        power = p
        while n % power == 0:
            for d in divisors:
                new_divisors.append(d * power)
            power *= p
        divisors.extend(new_divisors)
    return sorted(set(divisors))


def mobius_function(limit: int, primes: List[int]) -> List[int]:
    """Compute Mobius function."""
    mu = [1] * (limit + 1)
    is_square_free = [True] * (limit + 1)
    for p in primes:
        if p > limit:
            break
        for j in range(p, limit + 1, p):
            mu[j] *= -1
            if j % (p * p) == 0:
                is_square_free[j] = False
    for i in range(limit + 1):
        if not is_square_free[i]:
            mu[i] = 0
    return mu


def sq(n: int) -> int:
    """Square."""
    return n * n


def tr(n: int) -> int:
    """Triangular number."""
    return n * (n + 1) // 2


def solve() -> int:
    """Solve Problem 510."""
    N = 10**9
    L = isqrt(N)

    primes = sieve_primes(L)
    mu = mobius_function(L, primes)

    ans = 0
    for g in range(1, isqrt(N) + 1):
        if sq(g) > N:
            break
        n = N // sq(g)
        for b in range(1, isqrt(n) + 1):
            if sq(b) > n:
                break
            b_factors = prime_factorization(b, primes)
            divisors = all_divisors(sq(b), set(b_factors.keys()))
            for d in divisors:
                a = d - b
                if a > 0 and a <= b:
                    c = b - sq(b) // d
                    ans += (
                        mu[g]
                        * sq(g)
                        * (sq(a) + sq(b) + sq(c))
                        * tr(n // sq(b))
                    )

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/510.c
/*
 * Project Euler Problem 510: Tangent Circles.
 * Sum r_A + r_B + r_C for all valid circle configurations with r_A <= r_B <= N.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef long long ll;

int isqrt_ll(ll n) {
    ll r = (ll)sqrt((double)n);
    while (r * r > n) r--;
    while ((r + 1) * (r + 1) <= n) r++;
    return (int)r;
}

int main() {
    ll N = 1000000000LL;
    int L = isqrt_ll(N);

    /* Sieve primes up to L */
    int *primes = NULL;
    int num_primes = 0;
    {
        char *is_prime = (char*)calloc(L + 1, 1);
        memset(is_prime, 1, L + 1);
        is_prime[0] = is_prime[1] = 0;
        for (int i = 2; (ll)i * i <= L; i++)
            if (is_prime[i])
                for (int j = i * i; j <= L; j += i)
                    is_prime[j] = 0;
        for (int i = 2; i <= L; i++)
            if (is_prime[i]) num_primes++;
        primes = (int*)malloc(num_primes * sizeof(int));
        int idx = 0;
        for (int i = 2; i <= L; i++)
            if (is_prime[i]) primes[idx++] = i;
        free(is_prime);
    }

    /* Mobius function up to L */
    int *mu = (int*)calloc(L + 1, sizeof(int));
    {
        char *is_square_free = (char*)calloc(L + 1, 1);
        memset(is_square_free, 1, L + 1);
        for (int i = 0; i <= L; i++) mu[i] = 1;
        for (int pi = 0; pi < num_primes; pi++) {
            int p = primes[pi];
            if (p > L) break;
            for (int j = p; j <= L; j += p) {
                mu[j] *= -1;
                if (j % ((ll)p * p) == 0)
                    is_square_free[j] = 0;
            }
        }
        for (int i = 0; i <= L; i++)
            if (!is_square_free[i]) mu[i] = 0;
        free(is_square_free);
    }

    ll ans = 0;

    for (int g = 1; (ll)g * g <= N; g++) {
        if (mu[g] == 0) continue;
        ll n = N / ((ll)g * g);

        for (int b = 1; (ll)b * b <= n; b++) {
            ll b2 = (ll)b * b;

            /* Find all divisors of b^2 using prime factorization of b */
            int divs[2000];
            int ndivs = 0;
            {
                /* factorize b */
                int bb = b;
                int pf[20], pe[20];
                int npf = 0;
                for (int pi = 0; pi < num_primes && (ll)primes[pi] * primes[pi] <= bb; pi++) {
                    if (bb % primes[pi] == 0) {
                        pf[npf] = primes[pi];
                        pe[npf] = 0;
                        while (bb % primes[pi] == 0) {
                            pe[npf]++;
                            bb /= primes[pi];
                        }
                        npf++;
                    }
                }
                if (bb > 1) {
                    pf[npf] = bb;
                    pe[npf] = 1;
                    npf++;
                }
                /* Generate divisors of b^2 (exponents doubled) */
                divs[0] = 1;
                ndivs = 1;
                for (int i = 0; i < npf; i++) {
                    int old_count = ndivs;
                    ll pp = 1;
                    for (int e = 0; e < 2 * pe[i]; e++) {
                        pp *= pf[i];
                        for (int j = 0; j < old_count; j++) {
                            divs[ndivs++] = (int)(divs[j] * pp);
                        }
                    }
                }
            }

            for (int di = 0; di < ndivs; di++) {
                int d = divs[di];
                int a = d - b;
                if (a > 0 && a <= b) {
                    int c = b - (int)(b2 / d);
                    ll tri = (ll)b * (b + 1) / 2;  /* Wait, need n / b^2 */
                    ll nb2 = n / b2;
                    ll tr_nb2 = nb2 * (nb2 + 1) / 2;
                    ans += (ll)mu[g] * ((ll)g * g) *
                           ((ll)a * a + (ll)b * b + (ll)c * c) * tr_nb2;
                }
            }
        }
    }

    printf("%lld\n", ans);

    free(primes);
    free(mu);
    return 0;
}
